Objects/listobject.c

```diff
@@ -655,16 +659,11 @@ list_item(PyObject *aa, Py_ssize_t i)
         PyErr_SetObject(PyExc_IndexError, &_Py_STR(list_err));
         return NULL;
     }
-    PyObject *item;
-    Py_BEGIN_CRITICAL_SECTION(a);
 #ifdef Py_GIL_DISABLED
-    if (!_Py_IsOwnedByCurrentThread((PyObject *)a) && !_PyObject_GC_IS_SHARED(a)) {
-        _PyObject_GC_SET_SHARED(a);
-    }
+    return list_get_item_ref(a, i);
```

Comment:
We need to handle the case where `list_get_item_ref` returns `NULL` (because `i` is out of bounds)

Reply:
Isn't that already handled by the [check](https://github.com/python/cpython/blob/74072a3ffc733e32159e694bcf7a2198f2db0d43/Objects/listobject.c#L654) in `list_item`? Or is this needed to handle the case where another thread mutates the list after we perform the check?

Reply:
@mpage
Looks like Sam is correct. 
`list_get_item_ref` returns `NULL` without any error message, so it should be handled.
Since we remove the lock from this PR, it can happen anytime.

Please let me know it there is any misunderstanding.

https://github.com/python/cpython/blob/30b4e9f9c42493136c58c56fee5553128bb32428/Objects/listobject.c#L366-L379